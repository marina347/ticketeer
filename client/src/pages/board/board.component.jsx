import React from "react";
import { connect } from "react-redux";

import { selectBoard } from "../../redux/board/board.selectors";
import LaneList from "../../components/lane-list/lane-list.component";
import BoardDetails from "../../components/board-details/board-details.component";
import "./board.styles.scss";

const BoardPage = ({ board }) => {
  return (
    <div className="board-page">
      <div className="board-page__group">
        <BoardDetails board={board} />
        <LaneList boardId={board._id} />
      </div>
    </div>
  );
};

const mapStateToProps = (state, ownProps) => {
  return { board: selectBoard(ownProps.match.params.boardId)(state) };
};

export default connect(mapStateToProps)(BoardPage);
