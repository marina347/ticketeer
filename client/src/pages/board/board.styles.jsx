import styled from "styled-components";

export const LaneListContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 200px;
  background-color: lightpink;
`;