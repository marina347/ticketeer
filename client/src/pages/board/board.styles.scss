.board-page {
  display: flex;
  flex-direction: column;
  flex: 1 0 100%;
  &__content{
    display: flex;
    flex-direction: column;
    flex: 1 0 100%;
  }
}
