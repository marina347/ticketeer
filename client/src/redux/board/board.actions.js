import { toast } from "react-toastify";
import React from "react";

import BoardTypes from "./board.types";
import EnvVariables from "../../env-variables";
import { removeError } from "../common";
import ErrorMessages from "../../utils/error-messages";
import FormButton from "../../components/form-button/form-button.component";

export const getBoardsStart = () => ({
  type: BoardTypes.GET_BOARDS_START,
});

export const getBoardsSuccess = (boards) => ({
  type: BoardTypes.GET_BOARDS_SUCCESS,
  boards,
});

export const getBoardsFailure = (error) => ({
  type: BoardTypes.GET_BOARDS_FAILURE,
  error,
});

export const addBoardStart = () => ({
  type: BoardTypes.ADD_BOARD_START,
});

export const addBoardSuccess = (board) => ({
  type: BoardTypes.ADD_BOARD_SUCCESS,
  board,
});

export const addBoardFailure = (error) => ({
  type: BoardTypes.ADD_BOARD_FAILURE,
  error,
});

export const getTokenStart = () => ({
  type: BoardTypes.GET_TOKEN_START,
});

export const getTokenSuccess = (token) => ({
  type: BoardTypes.GET_TOKEN_SUCCESS,
  token,
});

export const getTokenFailure = (error) => ({
  type: BoardTypes.GET_TOKEN_FAILURE,
  error,
});

export const generateHashedBoardIdStart = () => ({
  type: BoardTypes.GENERATE_HASHED_BOARD_ID_START,
});

export const generateHashedBoardIdSuccess = (boardId, link) => ({
  type: BoardTypes.GENERATE_HASHED_BOARD_ID_SUCCESS,
  boardId,
  link,
});

export const generateHashedBoardIdFailure = (error) => ({
  type: BoardTypes.GENERATE_HASHED_BOARD_ID_FAILURE,
  error,
});

export const joinBoardStart = () => ({
  type: BoardTypes.JOIN_BOARD_START,
});

export const joinBoardSuccess = (board) => ({
  type: BoardTypes.JOIN_BOARD_SUCCESS,
  board,
});

export const joinBoardFailure = (error) => ({
  type: BoardTypes.JOIN_BOARD_FAILURE,
  error,
});

export const fetchBoardMembersStart = () => ({
  type: BoardTypes.FETCH_BOARD_MEMBERS_START,
});

export const fetchBoardMembersSuccess = (boardId, members) => ({
  type: BoardTypes.FETCH_BOARD_MEMBERS_SUCCESS,
  boardId,
  members,
});

export const fetchBoardMembersFailure = (error) => ({
  type: BoardTypes.FETCH_BOARD_MEMBERS_FAILURE,
  error,
});

export const getBoardsAsync = (token) => {
  return async (dispatch) => {
    dispatch(getBoardsStart());
    try {
      const requestOptions = {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
      };
      const response = await fetch(
        `${EnvVariables.REACT_APP_SERVER_PATH}/boards`,
        requestOptions
      );
      const responseJson = await response.json();
      dispatch(getBoardsSuccess(responseJson.boards));
    } catch (error) {
      dispatch(getBoardsFailure(error.message));
    }
  };
};

export const addBoardAsync = (boardItem, token) => {
  return async (dispatch) => {
    dispatch(addBoardStart());
    try {
      const requestOptions = {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({ ...boardItem }),
      };
      const response = await fetch(
        `${EnvVariables.REACT_APP_SERVER_PATH}/boards`,
        requestOptions
      );
      const board = await response.json();
      dispatch(addBoardSuccess(board));
    } catch (error) {
      dispatch(addBoardFailure(error.message));
      toast.error(ErrorMessages.ADD_BOARD_ERROR_MESSAGE, {
        autoClose: false,
        closeButton: <FormButton className="btn btn-notification" onClick={() => dispatch(removeError())}>X</FormButton>,
      });
    }
  };
};

export const generateHashedBoardIdAsync = (boardId, token) => {
  return async (dispatch) => {
    dispatch(generateHashedBoardIdStart());
    try {
      const requestOptions = {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
      };
      const response = await fetch(
        `${EnvVariables.REACT_APP_SERVER_PATH}/boards/make-invite/${boardId}`,
        requestOptions
      );
      const jsonResponse = await response.json();
      dispatch(
        generateHashedBoardIdSuccess(boardId, jsonResponse.hashedBoardId)
      );
    } catch (error) {
      dispatch(generateHashedBoardIdFailure(error.message));
      toast.error(ErrorMessages.GENERATE_INVITATION_LINK_ERROR_MESSAGE, {
        autoClose: false,
        closeButton: <FormButton className="btn btn-notification" onClick={() => dispatch(removeError())}>X</FormButton>,
      });
    }
  };
};

export const joinBoardAsync = (hashedBoardId, token) => {
  return async (dispatch) => {
    dispatch(joinBoardStart());
    try {
      const requestOptions = {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
      };
      const response = await fetch(
        `${EnvVariables.REACT_APP_SERVER_PATH}/boards/join-board/${hashedBoardId}`,
        requestOptions
      );
      const jsonResponse = await response.json();
      dispatch(joinBoardSuccess(jsonResponse.board));
    } catch (error) {
      dispatch(joinBoardFailure(error.message));
      toast.error(ErrorMessages.JOIN_BOARD_ERROR_MESSAGE, {
        autoClose: false,
        closeButton: <FormButton className="btn btn-notification" onClick={() => dispatch(removeError())}>X</FormButton>,
      });
    }
  };
};

export const fetchBoardMembersAsync = (boardId, token) => {
  return async (dispatch) => {
    dispatch(fetchBoardMembersStart());
    try {
      const requestOptions = {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
      };
      const response = await fetch(
        `${EnvVariables.REACT_APP_SERVER_PATH}/boards/${boardId}/members`,
        requestOptions
      );
      const jsonResponse = await response.json();
      dispatch(fetchBoardMembersSuccess(boardId, jsonResponse.members));
    } catch (error) {
      dispatch(fetchBoardMembersFailure(error.message));
    }
  };
};
