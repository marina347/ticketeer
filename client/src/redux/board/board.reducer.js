import BoardTypes from "./board.types";

const INITIAL_STATE = {
  boardItems: [],
  isLoading: false,
  error: null,
};

const boards = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case BoardTypes.GET_BOARDS:
      return { ...state, boardItems: action.boards };
    case BoardTypes.ADD_BOARD_START:
    case BoardTypes.GET_BOARDS_START:
    case BoardTypes.GENERATE_HASHED_BOARD_ID_START:
    case BoardTypes.JOIN_BOARD_START:
    case BoardTypes.FETCH_BOARD_MEMBERS_START:
      return { ...state, isLoading: true };
    case BoardTypes.ADD_BOARD_SUCCESS:
      return {
        ...state,
        isLoading: false,
        error: null,
        boardItems: [...state.boardItems, { ...action.board }],
      };
    case BoardTypes.GET_BOARDS_SUCCESS:
      return {
        ...state,
        isLoading: false,
        error: null,
        boardItems: action.boards,
      };
    case BoardTypes.GENERATE_HASHED_BOARD_ID_SUCCESS:
      return {
        ...state,
        isLoading: false,
        error: null,
        boardItems: state.boardItems.map((board) =>
          board._id == action.boardId ? { ...board, link: action.link } : board
        ),
      };
    case BoardTypes.JOIN_BOARD_SUCCESS:
      return {
        ...state,
        isLoading: false,
        error: null,
        boardItems: [...state.boardItems, action.board],
      };
    case BoardTypes.FETCH_BOARD_MEMBERS_SUCCESS:
      return {
        ...state,
        error: null,
        isLoading: false,
        boardItems: state.boardItems.map((board) =>
          board._id == action.boardId
            ? { ...board, members: action.members }
            : board
        ),
      };
    case BoardTypes.ADD_BOARD_FAILURE:
    case BoardTypes.GET_BOARDS_FAILURE:
    case BoardTypes.GENERATE_HASHED_BOARD_ID_FAILURE:
    case BoardTypes.JOIN_BOARD_FAILURE:
    case BoardTypes.FETCH_BOARD_MEMBERS_FAILURE:
      return { ...state, isLoading: false, error: action.error };
    case BoardTypes.MOVE_TICKET:
      return state.map((board) => helper(board, board.tasks, action));
    default:
      return state;
  }
};

const helper = (board, tickets, action) => {
  return {
    ...board,
    tasks: tickets.map((ticket) =>
      ticket.id === action.ticketId
        ? { ...ticket, type: action.ticketType }
        : ticket
    ),
  };
};

export default boards;
