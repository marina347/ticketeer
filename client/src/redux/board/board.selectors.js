import { createSelector } from "reselect";

const selectBoards = (state) => state.boards;

const selectMembers = (state) => state.user.members;

export const selectBoardsItems = createSelector(
  [selectBoards],
  (boards) => boards.boardItems
);

export const selectBoardsAreLoading = createSelector(
  [selectBoards],
  (boards) => boards.isLoading
);

export const selectBoard = (boardId) => {
  return createSelector([selectBoardsItems], (boards) =>
    boards.find((board) => board._id === boardId)
  );
};

export const selectBoardLink = (boardId) => {
  return createSelector([selectBoard(boardId)], (board) =>
    board.link ? board.link : ""
  );
};

export const selectMembersOfBoard = (boardId) => {
  return createSelector([selectBoard(boardId)], (board) => board.members);
};

export const selectMemberOfBoardByMemberId = (boardId, memberId) => {
  return createSelector([selectMembers], (members) =>
    members.find((member) => member._id == memberId)
  );
};

export const selectTasksFromBoard = (boardId) =>
  createSelector([selectBoard(boardId)], (board) => board.tasks);

export const selectTaskByStatus = (boardId, status) =>
  createSelector([selectTasksFromBoard(boardId)], (tasks) =>
    tasks.filter((task) => task.type === status)
  );

export const selectMemberObjectFromBoard = (boardId, userMembers) => {
  const selector = createSelector([selectMembersOfBoard(boardId)], (members) =>
    members.map((member) =>
      userMembers.find((userMember) => userMember._id == member.member)
    )
  );
  return selector;
};
