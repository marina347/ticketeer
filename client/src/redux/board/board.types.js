const BoardActions = {
  GET_BOARDS_START: "GET_BOARDS_START",
  GET_BOARDS_SUCCESS: "GET_BOARDS_SUCCESS",
  GET_BOARDS_FAILURE: "GET_BOARDS_FAILURE",

  ADD_BOARD_START: "ADD_BOARD_START",
  ADD_BOARD_SUCCESS: "ADD_BOARD_SUCCESS",
  ADD_BOARD_FAILURE: "ADD_BOARD_FAILURE",

  GET_BOARD_LINES: "GET_BOARD_LINES",
  GET_BOARD_SUCCESS: "GET_BOARD_SUCCESS",
  GET_BOARD_FAILURE: "GET_BOARD_FAILURE",

  GENERATE_HASHED_BOARD_ID_START: "ENERATE_HASHED_BOARD_ID_START",
  GENERATE_HASHED_BOARD_ID_SUCCESS: "ENERATE_HASHED_BOARD_ID_SUCCESS",
  GENERATE_HASHED_BOARD_ID_FAILURE: "ENERATE_HASHED_BOARD_ID_FAILURE",

  JOIN_BOARD_START: "JOIN_BOARD_START",
  JOIN_BOARD_SUCCESS: "JOIN_BOARD_SUCCES",
  JOIN_BOARD_FAILURE: "JOIN_BOARD_FAILURE",

  FETCH_BOARD_MEMBERS_START: "FETCH_BOARD_MEMBERS_START",
  FETCH_BOARD_MEMBERS_SUCCESS: "FETCH_BOARD_MEMBERS_SUCCESS",
  FETCH_BOARD_MEMBERS_FAILURE: "FETCH_BOARD_MEMBERS_FAILURE",
};

export default BoardActions;
