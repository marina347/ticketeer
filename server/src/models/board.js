const mongoose = require("mongoose");

const fkHelper = require("../utils/FKHelper");
const Lane = require("../models/lane");

const boardSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true,
  },
  description: {
    type: String,
  },
  creator: {
    type: mongoose.Schema.Types.ObjectId,
    required: true,
    ref: "User",
    validate: {
      async validator(value) {
        return await fkHelper.checkIfValueExistsInDatabase(
          mongoose.model("User"),
          value
        );
      },
      message: "User does not exists",
    },
  },
  members: [
    {
      member: {
        type: mongoose.Schema.Types.ObjectId,
        required: true,
        ref: "User",
        validate: {
          async validator(value) {
            return await fkHelper.checkIfValueExistsInDatabase(
              mongoose.model("User"),
              value
            );
          },
          message: "User does not exists",
        },
      },
    },
  ],
});

boardSchema.virtual("lanes", {
  ref: "Lane",
  localField: "_id",
  foreignField: "boardId",
});

boardSchema.pre("findOneAndDelete", async function (next) {
  const board = this;
  await Lane.deleteMany({ boardId: board._conditions._id });
  next();
});

const Board = mongoose.model("Board", boardSchema);
module.exports = Board;
