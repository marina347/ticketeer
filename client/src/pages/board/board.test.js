import React from "react";
import { shallow } from "enzyme";
import { BoardPage } from "./board.component";

describe("Board page", () => {
  let wrapper;
  let board;

  beforeEach(() => {
    board = {
      _id: "1",
    };
    wrapper = shallow(<BoardPage board={board} />);
  });

  it("Should render Board page", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
