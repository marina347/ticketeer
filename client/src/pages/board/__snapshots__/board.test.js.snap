// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board page Should render Board page 1`] = `
<div
  className="board-page"
>
  <div
    className="board-page__content"
  >
    <BoardDetails
      board={
        Object {
          "_id": "1",
        }
      }
    />
    <Connect(LaneList)
      boardId="1"
    />
  </div>
</div>
`;
